docker/docker_client.go

```diff
@@ -446,8 +399,9 @@ func (c *dockerClient) ping() (*pingResponse, error) {
 		if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusUnauthorized {
 			return nil, errors.Errorf("error pinging repository, response code %d", resp.StatusCode)
 		}
+		c.challenges = parseAuthHeader(resp.Header)
```

Comment:
This feels unclean; other values are only set via the `pr` structure, and `c.challenges` differs for no obvious reason.

(We could probably get rid of `pingResponse` altogether, have the nested function return a `(*http.Response, scheme string, error)`, and move the `resp.Header.Get()` calls and the data transfers which now happen at the top of `makeRequest` to the successful termination path of `ping`. Not sure whether _that_ would be the cleanest way, it’s just one of many options. Just adding `challenges` to `pingResponse` would also work. Or dropping `pingReponse` and making all the `c.` assignments in here.)